services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java

```diff
@@ -0,0 +1,283 @@
+            }
+
+            String oldClientSessionHost = clientSession.getNote(AdapterConstants.CLIENT_SESSION_HOST);
+            if (!Objects.equals(adapterSessionHost, oldClientSessionHost)) {
+                clientSession.setNote(AdapterConstants.CLIENT_SESSION_HOST, adapterSessionHost);
+            }
+        }
+    }
+
+    protected void updateUserSessionFromClientAuth(UserSessionModel userSession) {
+        for (Map.Entry<String, String> attr : clientAuthAttributes.entrySet()) {
+            userSession.setNote(attr.getKey(), attr.getValue());
+        }
+    }
+
+    protected void checkAndRetrieveDPoPProof(boolean isDPoPSupported) {
+        if (!isDPoPSupported) return;
+
+        if (clientConfig.isUseDPoP() || request.getHttpHeaders().getHeaderString(DPoPUtil.DPOP_HTTP_HEADER) != null) {
+            try {
+                dPoP = new DPoPUtil.Validator(session).request(request).uriInfo(session.getContext().getUri()).validate();
+                session.setAttribute(DPoPUtil.DPOP_SESSION_ATTRIBUTE, dPoP);
+            } catch (VerificationException ex) {
+                event.error(Errors.INVALID_DPOP_PROOF);
+                throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_DPOP_PROOF, ex.getMessage(), Response.Status.BAD_REQUEST);
+            }
+        }
+    }
+
+    protected String getRequestedScopes() {
+        String scope = formParams.getFirst(OAuth2Constants.SCOPE);
+
+        boolean validScopes;
+        if (Profile.isFeatureEnabled(Profile.Feature.DYNAMIC_SCOPES)) {
+            AuthorizationRequestContext authorizationRequestContext = AuthorizationContextUtil.getAuthorizationRequestContextFromScopes(session, scope);
+            validScopes = TokenManager.isValidScope(scope, authorizationRequestContext, client);
+        } else {
+            validScopes = TokenManager.isValidScope(scope, client);
+        }
+
+        if (!validScopes) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_SCOPE, "Invalid scopes: " + scope,
+                    Response.Status.BAD_REQUEST);
+        }
+
+        return scope;
+    }
+
+    protected void checkClient() {
+        AuthorizeClientUtil.ClientAuthResult clientAuth = AuthorizeClientUtil.authorizeClient(session, event, cors);
+        client = clientAuth.getClient();
+        clientAuthAttributes = clientAuth.getClientAuthAttributes();
+        clientConfig = OIDCAdvancedConfigWrapper.fromClientModel(client);
+
+        cors.allowedOrigins(session, client);
+
+        if (client.isBearerOnly()) {
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_CLIENT, "Bearer-only not allowed", Response.Status.BAD_REQUEST);
+        }
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+        OAuth2GrantManager.register(this);
```

Comment:
Related to the commend from @mposolda here https://github.com/keycloak/keycloak/pull/26251/files#r1472472745.

This is kinda against how providers/factories should be installed/inited. Now in addition to the usual installation of providers/factories (when the session factory is created and in Quarkus during re-aug) you are introducing another form of installing providers.

We should just rely on the existing mechanisms.

Another problem I see with this is the possibility of arbitrary extensions suddenly registering providers at runtime.